livemaker/cli/lmlsb.py

```diff
@@ -1159,3 +1159,154 @@ def scan_menus(lsb, csv_data, file_name, extract=False, patch=False):
+
+    if not no_backup:
+        print("Backing up original LSB.")
+        shutil.copyfile(str(lsb_file), "{}.bak".format(str(lsb_file)))
+    try:
+        new_lsb_data = lsb.to_lsb()
+        with open(lsb_file, "wb") as f:
+            f.write(new_lsb_data)
+        print("Wrote new LSB.")
+    except LiveMakerException as e:
+        sys.exit('Could not generate new LSB file: {}'.format(e))
+
+
+def scan_texts(lsb, csv_data, file_name, extract=False, patch=False):
+    """Scan given lsb object for TextIns, extract or patch TWdChar blocks
+    lsb: the lsb object loaded by LMScript.from_lsb
+    csv_data: translation data from the csv file
+    file_name: name of the lsb file
+    extract: true if you want to extract text
+    patch: true if you want to patch translated text
+    """
+    patched = False
+    for cmd_nr, cmd in enumerate(lsb.commands):
+        if cmd.type == CommandType.TextIns:
+            scenario = cmd.get('Text')
+            text_line = None
+            block_nr = 0
+            new_body = []
+            last_ch = None
+            cmd_patched = False
+            for w in scenario.body:
+                if w.type == TWdType.TWdChar:
+                    if text_line:
+                        text_line = text_line + str(w.ch)
+                    else:
+                        text_line = str(w.ch)
+                        block_nr += 1
+                    last_ch = w
+                else:
+                    if text_line:
+                        if extract:
+                            csv_data.append([file_name, cmd_nr, block_nr, text_line, None])
+
+                        if patch:
+                            for csv_line in csv_data:
+                                if csv_line[0] == file_name and csv_line[1] == str(cmd_nr) and csv_line[2] == str(block_nr):
+                                    if csv_line[4] == "":
+                                        print("Untranslated text \"{}\" at Command {}.{}! Not patched!".format(text_line, cmd_nr, block_nr))
+                                        new_line = text_line
+                                    else:
+                                        print("Patched \"{}\" to \"{}\" at Command {}.{}".format(text_line, csv_line[4], cmd_nr, block_nr))
+                                        new_line = csv_line[4]
+                                        cmd_patched = True
+                                    for ch in new_line:
+                                        new_ch = TWdChar(
+                                            ch = ch,
+                                            decorator = last_ch.decorator,
+                                            text_speed = last_ch.text_speed,
+                                            link_name = last_ch.link_name,
+                                            link = last_ch.link,
+                                            condition = last_ch.condition,
+                                        )
```

Comment:
This command probably needs some documented warning that we ill break cases where a game uses styling tags (decorators) or links/conditions. (but links/conditions aren't as important here since I think they are really only used in history/scrollback).

Ex:
Say you have a text line that looks like
```
<STYLE ID=1>おはよう</STYLE><STYLE ID=2>ございます</STYLE>
```
where style 1 is the default, and style 2 is text colored red.

the string is extracted as
```
おはようございます
```
but when we translate and re-insert it, it would end up looking like
```
<STYLE ID=2>Good morning</STYLE>
```
where the entire line is now red, since the decorator condition is coming from the last character in the line.

It's not a huge deal, but basically any one using this command/API needs to aware that if their game uses tags for more advanced formatting, they will still need to clean things up themselves using the full script format.

and in the event a game is actually using links/conditions for embedding icons/images into regular message boxes, it may break their game (until they fix the translated script).

Reply:
I am not sure about this. The script does a line break on everything else than TWdChar.
As long links/conds/icons/images are not embedded into TWdChar they remain an the right places.
If they are embedded you are right.

Reply:
Yeah the issue is that links/conditionals/decorators are flags tied to a TWdChar. when a decorator (font styling) changes between characters, the only difference is that the next TWdChar will have a new `.decorator` value (corresponding to some font style ID). So there's no non-TWdChar object between the different characters.

if you look at what the script decompiler does, it has to watch all of those flags for whenever they change, and then close any previously opened style/link/etc tag, and then open a tag for the new style/link/etc.

Reply:
It's probably not a serious issue for a majority of LM games out there, but it's still something people should be aware of.